### PR TITLE
[80375] Enable filtering and sorting for model zoo

### PR DESCRIPTION
**Pull request by yatarkan**
### All Submissions:
* [x] Create a meaningful name of the PR with reference to JIRA. Example: *[43714] [NIGHTLY] Take correct tag for snyk scans*
* [x] JIRA tickets:
  * 80375 - OMZ page filtering and sorting panel
* [x] Set assignee to the PR.
* [x] Add corresponding labels to the PR. 
* [x] Set appropriate milestone to the PR.
* [ ] Added tests for new functionality (unit / e2e)
* [x] Followed code style guidelines:
  * Code covered with typings
  * No debug printing in production code
  * No commented blocks of code in production code

### CODE REVIEW

File: client/src/app/modules/model-manager/components/omz-import-ribbon-content/omz-import-ribbon-content.component.ts [hunk 1]
```diff
@@ -116,6 +149,38 @@ export class OmzImportRibbonContentComponent implements AfterViewInit, OnDestroy
...
+    return this._getOptionsByKey('task_type');
+  }
+
+  get precisionOptions(): string[] {
+    return this._getOptionsByKey('precision');
+  }
+
+  get frameworkOptions(): string[] {
+    return this._getOptionsByKey('framework');
+  }
```
Comment (as-suvorov) [inline client/src/app/modules/model-manager/components/omz-import-ribbon-content/omz-import-ribbon-content.component.ts:175]: Due to the time complexity and change detection cycles it makes sense to move options calculations out of getters.

File: client/src/app/shared/models/model-zoo-data-source/base-model-zoo-data-source.ts [hunk 2]
```diff
@@ -150,11 +158,11 @@ export abstract class BaseModelZooDataSource<T, U = string> implements DataSourc
   }
 
   // reuse mat table data source implementation to handle edge cases
-  private _sortData(data: T[], sort: IModelZooSort): T[] {
+  private _sortData(data: T[], sort: IModelZooSort<T>): T[] {
     if (!sort) {
       return data;
     }
-    const active = sort.active;
+    const active = sort.field as string;
```
Comment (as-suvorov) [inline client/src/app/shared/models/model-zoo-data-source/base-model-zoo-data-source.ts:165]: We need to handle this case gracefully as according to `IModelZooSort` interface non `string` is a valid `field` type.

File: client/src/app/shared/models/model-zoo-data-source/huggingface-model-zoo-data-source.ts [hunk 3]
```diff
@@ -11,9 +11,16 @@ export class HuggingfaceModelZooDataSource extends BaseModelZooDataSource<
   IHuggingfaceModel,
   IHuggingfaceModelZooFilter
 > {
+  sortOptions: IModelZooSort<IHuggingfaceModel>[] = [
```
Comment (as-suvorov) [inline client/src/app/shared/models/model-zoo-data-source/huggingface-model-zoo-data-source.ts:14]: `readonly`?)

**Review by as-suvorov — CHANGES_REQUESTED** [review]

File: client/src/app/shared/models/model-zoo-data-source/huggingface-model-zoo-data-source.ts (hunk 3 above)
Comment (yatarkan) [reply client/src/app/shared/models/model-zoo-data-source/huggingface-model-zoo-data-source.ts:14]: Fixed

File: client/src/app/modules/model-manager/components/hugging-face-import-ribbon-content/hugging-face-import-ribbon-content.component.ts [hunk 4]
```diff
@@ -43,22 +44,30 @@ export interface IHuggingfaceTagsSets {
...
 
-  readonly dataSource = new HuggingfaceModelZooDataSource();
-
   selectedModel: IHuggingfaceModel = null;
 
   @ViewChild(MatPaginator) paginator: MatPaginator;
 
+  readonly appliedFiltersCount$ = this.filterControl.valueChanges.pipe(
+    map(
+      (filters: Record<keyof ModelDownloaderDTO, string[]>) =>
```
[the remark this answers is not in the document]
Comment (yatarkan) [reply client/src/app/modules/model-manager/components/hugging-face-import-ribbon-content/hugging-face-import-ribbon-content.component.ts:66]: Copied from OMZ component, missed it, thanks
Fixed

File: client/src/app/modules/model-manager/components/omz-import-ribbon-content/omz-import-ribbon-content.component.ts (hunk 1 above)
Comment (yatarkan) [reply client/src/app/modules/model-manager/components/omz-import-ribbon-content/omz-import-ribbon-content.component.ts:175]: Fixed

**Review by as-suvorov — APPROVED** [review]

File: client/src/app/modules/model-manager/components/omz-import-ribbon-content/omz-import-ribbon-content.component.html [hunk 5]
```diff
@@ -1,5 +1,21 @@
...
+    </mat-form-field>
+
+    <wb-model-zoo-filters-header
+      [appliedFiltersCount]="appliedFiltersCount$ | async"
+      (resetAllFilters)="resetAllFilters()"
+    ></wb-model-zoo-filters-header>
+
+    <ng-container ngProjectAs="wb-model-zoo-filter-group" [ngTemplateOutlet]="filterGroupTemplate"></ng-container>
+
     <wb-search-field ngProjectAs="model-zoo-search" (search)="searchModels($event)"></wb-search-field>
```
[the remark this answers is not in the document]
Comment (yatarkan) [reply client/src/app/modules/model-manager/components/omz-import-ribbon-content/omz-import-ribbon-content.component.html:19]: Discussed. Will be addressed in next PR

File: client/src/app/modules/model-manager/components/model-zoo-filter-group/model-zoo-filter/model-zoo-filter-option/model-zoo-filter-option.component.ts [hunk 6]
```diff
@@ -21,15 +22,19 @@ export interface IModelZooFilterOptionChangeEvent<T> {
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class ModelZooFilterOptionComponent<T> {
+  @Input() label: string;
```
[the remark this answers is not in the document]
Comment (yatarkan) [reply client/src/app/modules/model-manager/components/model-zoo-filter-group/model-zoo-filter/model-zoo-filter-option/model-zoo-filter-option.component.ts]: Fixed - moved to content projection

**Review by vishniakov-nikolai — APPROVED** [review]

File: client/src/app/modules/model-manager/components/hugging-face-import-ribbon-content/hugging-face-import-ribbon-content.component.ts [hunk 7]
```diff
@@ -85,36 +73,16 @@ export class HuggingFaceImportRibbonContentComponent implements OnInit, AfterVie
...
-
-  ngOnDestroy(): void {
-    this._unsubscribe$.next();
-    this._unsubscribe$.complete();
-  }
-
-  searchModels(value: string): void {
-    this.idSearch = value;
-    this._filter();
+  protected get _dataSourceFilter(): IHuggingfaceModelZooFilter {
```
[the remark this answers is not in the document]
Comment (yatarkan) [reply client/src/app/modules/model-manager/components/hugging-face-import-ribbon-content/hugging-face-import-ribbon-content.component.ts:78]: It is protected as it is inherited from base abstract class

**Pull request merged**